src/kubernetes_charm.py

```diff
@@ -83,8 +100,126 @@ def _upgrade(self) -> typing.Optional[kubernetes_upgrade.Upgrade]:
+                res=lightkube.resources.core_v1.Service,
+                name=self._service_name,
+                namespace=self.model.name,
+            )
+        except lightkube.core.exceptions.ApiError as e:
+            if e.status.code == 404:
+                return None
+            raise
+
+        return service
+
+    @functools.cache
+    def _get_pod(self, unit_name: str) -> lightkube.resources.core_v1.Pod:
+        """Get the pod for the provided unit name."""
+        return self._lightkube_client.get(
+            res=lightkube.resources.core_v1.Pod,
+            name=unit_name.replace("/", "-"),
+            namespace=self.model.name,
+        )
+
+    @functools.cache
+    def _get_node(self, unit_name: str) -> lightkube.resources.core_v1.Node:
+        """Return the node for the provided unit name."""
+        node_name = self._get_pod(unit_name).spec.nodeName
+        return self._lightkube_client.get(
+            res=lightkube.resources.core_v1.Node,
+            name=node_name,
+            namespace=self.model.name,
+        )
+
+    def _apply_service(self, service_type: ServiceType, service_exists: bool = False) -> None:
+        """Apply the service type provided."""
+        pod0 = self._get_pod(f"{self.app.name}/0")
+
+        service = lightkube.resources.core_v1.Service(
+            metadata=lightkube.models.meta_v1.ObjectMeta(
+                name=self._service_name,
+                namespace=self.model.name,
+                ownerReferences=pod0.metadata.ownerReferences,  # the stateful set
+                labels={"app.kubernetes.io/name": self.app.name},
+            ),
+            spec=lightkube.models.core_v1.ServiceSpec(
+                ports=[
+                    lightkube.models.core_v1.ServicePort(
+                        name="mysql-rw",
+                        port=self._READ_WRITE_PORT,
+                        targetPort=self._READ_WRITE_PORT,
+                    ),
+                    lightkube.models.core_v1.ServicePort(
+                        name="mysql-ro",
+                        port=self._READ_ONLY_PORT,
+                        targetPort=self._READ_ONLY_PORT,
+                    ),
+                ],
+                type=service_type.value,
+                selector={"app.kubernetes.io/name": self.app.name},
+            ),
+        )
+
+        # Delete and re-create until https://bugs.launchpad.net/juju/+bug/2084711 resolved
+        if service_exists:
+            logger.info(f"Deleting service {service_type=}")
+            self._lightkube_client.delete(
+                res=lightkube.resources.core_v1.Service,
+                name=self._service_name,
+                namespace=self.model.name,
+            )
+            logger.info(f"Deleted service {service_type=}")
+
+        logger.info(f"Applying service {service_type=}")
+        self._lightkube_client.apply(service, field_manager=self.app.name)
+        logger.info(f"Applied service {service_type=}")
```

Comment:
filed a juju bug: https://bugs.launchpad.net/juju/+bug/2084711, havent heard back yet

essentially, we have included deletion + recreation of service as a workaround until we get help from juju to determine what may be happening

Reply:
no, i did not yet test with EKS or GKE. i dont believe that testing with these platforms should necessarily be a blocker for this PR

we tested on AKS, and this issue did not manifest itself in AKS

Reply:
i agree, but we are unable to run integration tests without deleting and recreating (as tests experience flicker of the juju client and fail with a `Bad file descriptor` error). @paulomach please share your thoughts when you are able

Reply:
As @shayancanonical , I validated the flaky behavior independently of a charm, but saw no issue in AKS.
Independently, this does not seems to be an issue in the charm/lightkube.
So let's not block the PR, since this can be refactored once we have better understanding or a fix.